### ISSUE TITLE
`Start-Process -Wait` always waits for a whole number of seconds on Windows

### ISSUE THREAD

**Issue by MatejKafka** [PowerShell/PowerShell#24709]
### Steps to reproduce

Run the following PowerShell snippet on Windows:
```powershell
Measure-Command {Start-Process cmd -ArgumentList "/c", "echo", "hello" -NoNewWindow -Wait} | select TotalMilliseconds
Measure-Command {Start-Process cmd -ArgumentList "/c", "echo", "hello" -NoNewWindow -PassThru | Wait-Process} | select TotalMilliseconds
```

### Expected behavior

Both commands finish almost instantly (~20 ms).

### Actual behavior

`Start-Process -Wait` always takes at least 1 second. `Start-Process | Wait-Process` is much quicker.

```console
TotalMilliseconds
-----------------
          1012.42
            15.02
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.5.0-rc.1
PSEdition                      Core
GitCommitId                    7.5.0-rc.1
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Root cause

`Start-Process -Wait` is currently implemented by polling once per second to check if the process exited:
https://github.com/PowerShell/PowerShell/blob/e12734582011d1f482daf4b669ad4dceb287828e/src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs#L2160-L2165

`WaitOne` is implemented here:
https://github.com/PowerShell/PowerShell/blob/e12734582011d1f482daf4b669ad4dceb287828e/src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs#L2806-L2813

Effectively, what the code does is to add the spawned process to a Win32 job object and then poll it every second to check if it exited. This is incredibly inefficient for short-running commands, and even for longer commands, the overhead from polling and the added delay is problematic.

`Wait-Process` is using a different implementation of waiting by adding an event handler for the `Exited` event on the awaited `Process` instance, which likely internally waits on the process handle and reacts immediately to the process exit, without polling.

I think `Start-Process` should use the same mechanism as `Wait-Process` to avoid the delay.

**Comment by 237dmitry** [comment]
```
TotalMilliseconds
-----------------
             9.63
            10.35
```

**Comment by MatejKafka** [comment]
@237dmitry Good point, the issue is Windows-specific – on Linux, a different waiting mechanism is used.

**Comment by jborean93** [comment]
I agree it should be fixed to avoid this problem, the `Start-Process -Wait` and `Wait-Process` implementation is different on Windows for a specific reason. Using `-Wait` on `Start-Process` will have it wait for the spawned process and any subsequent child processes it spawns to finish whereas `Wait-Process` just waits on the one process specified. This is the reason why the job object is used so it can encapsulate all child processes spawned and wait for them before continuing.

**Comment by MatejKafka** [comment]
@jborean93 Thanks, that makes sense, although imo it's a weird default to wait for all processes, not just the direct child process.

If we want to keep this behavior, the solution is a bit more elaborate, but still doable (and I suspect the native .NET thread-pool should be able to do some of the heavy lifting, although I'd have to ask someone on the .NET team about that): https://devblogs.microsoft.com/oldnewthing/20130405-00/?p=4743

**Comment by MatejKafka** [comment]
@jborean93 But it seems you know more about that part than I do. :)
https://github.com/jborean93/ProcessEx/blob/437e3bd350dcaf069968f564a5c083f4261148da/src/ProcessEx/Process.cs#L855

**Comment by jborean93** [comment]
Yep I started on a local branch with those changes but didn’t finish it off yet. Will open a PR when I do.

**Comment by jborean93** [comment]
I've opened https://github.com/PowerShell/PowerShell/pull/24711 which uses the same mechanism referenced by the blog post and what I use in my `ProcessEx` module.

**Comment by iSazonov** [comment]
@jborean93 I know we don't support Windows 7, nevertheless users can ask will it work on Windows 7? Do you know another side affects or possible breaking changes?

**Comment by jborean93** [comment]
Does PowerShell even launch on Windows 7 anymore? I recall seeing issues saying it fails to launch due to dlls that don't exist there anymore. In saying that AFAIK all the APIs added work back to Windows XP, I can't say anything about using the API set dlls though, I believe they are a Windows 8/Server 2012 feature.

I am not aware of other side affects or other breaking changes, the blog post originally shared is why I wrote ProcessEx the way it is and I haven't had any problems there https://devblogs.microsoft.com/oldnewthing/20130405-00/?p=4743. In saying that there could be an unknown side effect but this seems to be the "official" way to wait for processes in a job to finish.

**Comment by iSazonov** [comment]
I'd want add more info in PR description so that users can better understand which behavior change is by-design and which is not.

**Comment by jborean93** [comment]
I've added more info but I don't think there are any breaking changes here, the PR improves the logic to no longer having to poll every second to something where it gets notified on process completion allowing it to return immediately rather than every second.
